File: baekjoon/22-07/2178.js
// 로컬에서 테스트
let fs = require("fs");
let input = fs.readFileSync("./2178.txt").toString().trim();
input = input.split("\n");

const sol = (input) => {
  const [N, M] = input[0].split(" ").map((v) => Number(v));
  const graph = [];
  for (let i = 1; i <= N; i++)
    graph.push(
      input[i]
        .replace(/\r/g, "")
        .split("")
        .map((v) => Number(v))
    );

  function bfs(x, y) {
    const dx = [-1, 1, 0, 0];
    const dy = [0, 0, -1, 1];
    const queue = [];
    queue.push([x, y]);
    while (queue.length) {
      const [x, y] = queue.shift();
      for (let i = 0; i < 4; i++) {
        const nx = x + dx[i];
        const ny = y + dy[i];
        if (nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
        if (graph[nx][ny] === 0) continue;
        if (graph[nx][ny] === 1) {
          graph[nx][ny] = graph[x][y] + 1;
          queue.push([nx, ny]);
        }
      }
    }
  }
  bfs(0, 0);
  return graph[N - 1][M - 1];
};
console.log(sol(input));
/*
// 백준 제출 코드
const input = [];
require("readline")
  .createInterface(process.stdin, process.stdout)
  .on("line", (line) => {
    input.push(line);
  })
  .on("close", () => {
    console.log(sol(input));
    process.exit();
  });
*/

File: baekjoon/22-07/2178.py
# 22-07-22 [1]
from collections import deque
import sys
sys.stdin = open('2178.txt')

input = sys.stdin.readline

n, m = map(int, input().rstrip().split())

graph = []
for _ in range(n):
    graph.append(list(map(int, sys.stdin.readline().rstrip())))


def bfs(x, y):
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]

    q = deque()
    q.append((x, y))
    while q:
        x, y = q.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue

            if graph[nx][ny] == 0:
                continue

            if graph[nx][ny] == 1:
                graph[nx][ny] = graph[x][y] + 1
                q.append((nx, ny))
    return graph[n-1][m-1]


print(bfs(0, 0))
